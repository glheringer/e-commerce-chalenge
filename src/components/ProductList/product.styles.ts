import styled from 'styled-components';

export const ProductContainer = styled.div`
   font-family: 'Heebo', sans-serif;
   display: flex;
   flex-wrap: wrap;
   justify-content: center;
   padding: 32px;
   gap: 16px;
   box-sizing: border-box;
   max-width: 1200px;
   margin: 0 auto;
`;

export const ProductItem = styled.div`
  background-color: #f9f9f9;
  padding: 16px;
  border-radius: 8px;
  width: calc(33.33% - 32px);
  box-sizing: border-box;
  margin-bottom: 16px;
  display: flex;
  flex-direction: column;
  font-family: 'Heebo', sans-serif;

  img {
    width: 100%;
    height: auto;
    object-fit: contain;
    margin-bottom: 8px;
  }

  h2 {
    font-family: 'Heebo', sans-serif;
    color: #000;
    font-weight: 400;
    font-size: 14px;
    font-style: normal;
    line-height: 150%;
    letter-spacing: .56px;

    padding-bottom: 8px;
    border-bottom: solid;
    border-width: 1px;
    border-color: #D1D1D1;
  }

  h3 {
    font-family: 'Heebo', sans-serif;
    font-style: normal;
    font-weight: 700;
    line-height: 150%;
    font-size: 14px;
    letter-spacing: .56px;
    margin-bottom: 8px;
  }

  p {
    font-family: 'Heebo', sans-serif;
    margin: 4px 0;
  }
  //Para exibir 2 produtos por linha
  @media (max-width: 768px) {
    width: calc(50% - 32px);
  }
  //Para exibir 1 produtos por linha
  @media (max-width: 480px) {
    width: calc(100% - 32px);
  }
`;

export const ImageContainer = styled.div`
  overflow: hidden;
`;

export const ProductImage = styled.img`
  width: 100%;
  height: 200px;
  object-fit: contain;
  margin-bottom: 8px;
  transition: transform 0.5s ease;

  &:hover {
    transform: scale(1.35);
  }
`;

export const PreviousPrice = styled.p`
  margin: 4px 0;
  text-decoration: line-through;
  font-family: 'Heebo', sans-serif;
`;
