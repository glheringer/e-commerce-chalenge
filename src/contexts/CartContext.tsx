import React, { createContext, useState, useContext, ReactNode } from 'react';

interface CartContextType {
  cartItems: any[];
  totalPrice: number;
  addToCart: (item: any) => void;
//   removeFromCart: (itemId: number) => void;
}


const CartContext = createContext<CartContextType | undefined>(undefined);

export const CartProvider = ({ children }: { children: ReactNode }) => {
  const [cartItems, setCartItems] = useState<any[]>([]);
  const [totalPrice, setTotalPrice] = useState(0);

  const addToCart = (item: any) => {
    setCartItems((prevItems) => [...prevItems, item]);
    setTotalPrice((prevTotal) => prevTotal + item.price);
  };

  // Função para remover um item do carrinho
//   const removeFromCart = (itemId: number) => {
//     setCartItems((prevItems) => prevItems.filter(item => item.id !== itemId));
//     const item = cartItems.find(item => item.id === itemId);
//     if (item) {
//       setTotalPrice((prevTotal) => prevTotal - item.price);
//     }
//   };

  return (
    <CartContext.Provider value={{ cartItems, totalPrice, addToCart }}>
      {children}
    </CartContext.Provider>
  );
};

export const useCart = () => {
  const context = useContext(CartContext);
  if (!context) {
    throw new Error('useCart deve ser usado dentro de um CartProvider');
  }
  return context;
};
