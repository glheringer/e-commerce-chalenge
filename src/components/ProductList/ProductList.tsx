import React from 'react';
import { useCart } from '../../contexts/CartContext';

import { ImageContainer, PreviousPrice, ProductContainer, ProductImage, ProductItem } from './product.styles';
import AddCartButton from '../AddCartButton/AddCartButton';

interface Product {
    listPrice: number;
    price: number;
    name: string | undefined;
    image: string | undefined;
 
  }

interface ProductsProps{
    products: Product[];
}
const ProductList = ({ products}:ProductsProps) => {
    const { addToCart } = useCart(); 

    const handleAddToCart = (product: Product) => {
        addToCart(product);
    };

    return (
        <ProductContainer>
            {products.map((product, index) => (
                <ProductItem key={index}>
                  <ImageContainer>
                        <ProductImage src={product.image} alt={product.name} />
                    </ImageContainer>
                    <h2>{`R$${product.price.toFixed(2)}`}</h2>
                    <h3>{product.name}</h3>
                    <PreviousPrice>{`De: $${product.listPrice.toFixed(2)}`}</PreviousPrice>
                    <p>{`Por: $${product.price.toFixed(2)}`}</p>
                    <AddCartButton onClick={() => handleAddToCart(product)}>
                        Adicionar ao Carrinho
                    </AddCartButton>
                </ProductItem>
            ))}
        </ProductContainer>
    );
};

export default ProductList;
