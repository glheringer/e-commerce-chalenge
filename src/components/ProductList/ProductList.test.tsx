import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import ProductList from './ProductList';
import { useCart } from '../../contexts/CartContext';
import '@testing-library/jest-dom';

jest.mock('../../contexts/CartContext', () => ({
  useCart: jest.fn(),
}));

describe('ProductList', () => {
  const mockProducts = [
    {
      listPrice: 120,
      price: 100,
      name: 'Produto 1',
      image: 'https://via.placeholder.com/150',
    },
    {
      listPrice: 250,
      price: 200,
      name: 'Produto 2',
      image: 'https://via.placeholder.com/150',
    },
  ];
  const mockAddToCart = jest.fn();

  beforeEach(() => {
      (useCart as jest.Mock).mockReturnValue({
          addToCart: mockAddToCart,
      });
  });
  
  it('should render correctly', () => {

    (useCart as jest.Mock).mockReturnValue({
      addToCart: jest.fn(),
    });

    const { container } = render(<ProductList products={mockProducts} />);
    
    expect(container).toMatchSnapshot();
  });

  it('should call addToCart when "Adicionar ao Carrinho" button is clicked', () => {
    const { container } = render(<ProductList products={mockProducts} />);
    
    expect(container).toMatchSnapshot();

    const addToCartButtons = screen.getAllByText('Adicionar ao Carrinho');
    expect(addToCartButtons).toHaveLength(mockProducts.length);

    fireEvent.click(addToCartButtons[0]);
    expect(mockAddToCart).toHaveBeenCalledWith(mockProducts[0]);

    fireEvent.click(addToCartButtons[1]);
    expect(mockAddToCart).toHaveBeenCalledWith(mockProducts[1]);
  });
  
  it('should display correct price and list price for each product', () => {
    const { getByText } = render(<ProductList products={mockProducts} />);
  
    expect(screen.getByText('R$100.00')).toBeInTheDocument(); 
    expect(screen.getByText('De: $120.00')).toBeInTheDocument();
  });
  
});

