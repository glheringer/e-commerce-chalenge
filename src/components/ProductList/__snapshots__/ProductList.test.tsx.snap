// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList should call addToCart when "Adicionar ao Carrinho" button is clicked 1`] = `
<div>
  <div
    class="sc-blHHSb bfVLgD"
  >
    <div
      class="sc-gtLWhw iEWsjm"
    >
      <div
        class="sc-egkSDF gJHYNM"
      >
        <img
          alt="Produto 1"
          class="sc-fAUdSK fHCJtw"
          src="https://via.placeholder.com/150"
        />
      </div>
      <h2>
        R$100.00
      </h2>
      <h3>
        Produto 1
      </h3>
      <p
        class="sc-dntaoT jiGTwa"
      >
        De: $120.00
      </p>
      <p>
        Por: $100.00
      </p>
      <button
        class="sc-ivxoEo htZkwW"
      >
        <div
          class="sc-ghWlax dNFRXS"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium sc-kLhKbu eMIYyl css-20bmp1-MuiSvgIcon-root"
            data-testid="ShoppingCartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2M1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2"
            />
          </svg>
        </div>
        Adicionar ao Carrinho
      </button>
    </div>
    <div
      class="sc-gtLWhw iEWsjm"
    >
      <div
        class="sc-egkSDF gJHYNM"
      >
        <img
          alt="Produto 2"
          class="sc-fAUdSK fHCJtw"
          src="https://via.placeholder.com/150"
        />
      </div>
      <h2>
        R$200.00
      </h2>
      <h3>
        Produto 2
      </h3>
      <p
        class="sc-dntaoT jiGTwa"
      >
        De: $250.00
      </p>
      <p>
        Por: $200.00
      </p>
      <button
        class="sc-ivxoEo htZkwW"
      >
        <div
          class="sc-ghWlax dNFRXS"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium sc-kLhKbu eMIYyl css-20bmp1-MuiSvgIcon-root"
            data-testid="ShoppingCartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2M1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2"
            />
          </svg>
        </div>
        Adicionar ao Carrinho
      </button>
    </div>
  </div>
</div>
`;

exports[`ProductList should render correctly 1`] = `
<div>
  <div
    class="sc-blHHSb bfVLgD"
  >
    <div
      class="sc-gtLWhw iEWsjm"
    >
      <div
        class="sc-egkSDF gJHYNM"
      >
        <img
          alt="Produto 1"
          class="sc-fAUdSK fHCJtw"
          src="https://via.placeholder.com/150"
        />
      </div>
      <h2>
        R$100.00
      </h2>
      <h3>
        Produto 1
      </h3>
      <p
        class="sc-dntaoT jiGTwa"
      >
        De: $120.00
      </p>
      <p>
        Por: $100.00
      </p>
      <button
        class="sc-ivxoEo htZkwW"
      >
        <div
          class="sc-ghWlax dNFRXS"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium sc-kLhKbu eMIYyl css-20bmp1-MuiSvgIcon-root"
            data-testid="ShoppingCartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2M1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2"
            />
          </svg>
        </div>
        Adicionar ao Carrinho
      </button>
    </div>
    <div
      class="sc-gtLWhw iEWsjm"
    >
      <div
        class="sc-egkSDF gJHYNM"
      >
        <img
          alt="Produto 2"
          class="sc-fAUdSK fHCJtw"
          src="https://via.placeholder.com/150"
        />
      </div>
      <h2>
        R$200.00
      </h2>
      <h3>
        Produto 2
      </h3>
      <p
        class="sc-dntaoT jiGTwa"
      >
        De: $250.00
      </p>
      <p>
        Por: $200.00
      </p>
      <button
        class="sc-ivxoEo htZkwW"
      >
        <div
          class="sc-ghWlax dNFRXS"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium sc-kLhKbu eMIYyl css-20bmp1-MuiSvgIcon-root"
            data-testid="ShoppingCartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2M1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2"
            />
          </svg>
        </div>
        Adicionar ao Carrinho
      </button>
    </div>
  </div>
</div>
`;
